graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -584,6 +628,305 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
+        # Construct mapping from location -> list of filter predicates applied at that location
+        location_to_predicates = _construct_location_to_predicate_list(match_query)
+
+        new_match_traversals = []
+        for match_traversal in match_query.match_traversals:
+            new_match_traversal = _collect_filters_to_first_location_in_match_traversal(
+                match_traversal, location_to_predicates)
+            new_match_traversals.append(new_match_traversal)
+
+        new_match_queries.append(
+            MatchQuery(
+                match_traversals=new_match_traversals,
+                folds=match_query.folds,
+                output_block=match_query.output_block
+            )
+        )
+
+    return CompoundMatchQuery(match_queries=new_match_queries)
+
+
+def _update_context_field_binary_composition(expression, present_locations):
+    """Lower BinaryCompositions involving non-existent ContextFields to True.
+
+    Args:
+        expression: BinaryComposition with at least one ContextField operand
+
+    Returns:
+        TrueLiteral iff either ContextField operand is not in `present_locations`,
+        and the original expression otherwise
+    """
+    if isinstance(expression.left, ContextField):
+        context_field = expression.left
+        location_name, _ = context_field.location.get_location_name()
+        if location_name not in present_locations:
+            return TrueLiteral
+
+    if isinstance(expression.right, ContextField):
+        context_field = expression.right
+        location_name, _ = context_field.location.get_location_name()
+        if location_name not in present_locations:
+            return TrueLiteral
+
+    return expression
+
+
+def _simplify_non_context_field_binary_composition(expression):
+    """Return a simplified BinaryComposition if either operand is a TrueLiteral.
+
+    Args:
+        expression: BinaryComposition without any ContextField operand(s)
+
+    Returns:
+        simplified expression if the given expression is a disjunction/conjunction
+        and one of it's operands is a TrueLiteral,
+        and the original expression otherwise
+    """
+    if expression.operator == u'||':
+        if expression.left == TrueLiteral or expression.right == TrueLiteral:
+            return TrueLiteral
+        else:
+            return expression
+    elif expression.operator == u'&&':
+        if expression.left == TrueLiteral:
+            return expression.right
+        if expression.right == TrueLiteral:
+            return expression.left
+        else:
+            return expression
+    else:
+        return expression
+
+
+def _simplify_ternary_conditional(expression):
+    """Return the `if_true` clause if the predicate of the TernaryConditional is a TrueLiteral.
+
+    Args:
+        expression: TernaryConditional to be simplified.
+
+    Returns:
+        simplified TernaryConditional, if the predicate is True,
```

Comment:
this is misleading -- if the predicate was `TrueLiteral`, the return isn't a `TernaryConditional` at all